File: P2/ruby/SpaceStation.rb
#encoding: UTF-8

require_relative 'SpaceStationToUI'

module Deepspace
    class SpaceStation
        @@MAXFUEL = 100
        @@SHIELDLOSSPERUNITSHOT = 0.1

        def initialize (n, supplies) # SpaceStation(n : string, supplies : SuppliesPackage)
			@name = n
			@ammoPower = supplies.ammoPower
            @fuelUnits = supplies.fuelUnits
			@shieldPower = supplies.shieldPower			
			
			@nMedals = 0
            @pendingDamage = nil # : Damage
            @weapons = Array.new # : Array<Weapon>
            @shieldBoosters = Array.new # : ShieldBooster
            @hangar = nil # : Hangar
		end
		
		attr_reader :name, :ammoPower, :fuelUnits, :shieldPower, :nMedals, :pendingDamage, :weapons, :shieldBoosters, :hangar

		def getUIversion
			SpaceStationToUI.new(self)
		end

		def to_s
			"MAXFUEL: #{@@MAXFUEL}, SHIELDLOSSPERUNITSHOT: #{@@SHIELDLOSSPERUNITSHOT}, ammoPower: #{@ammoPower}, fuelUnits: #{@fuelUnits}, name: #{@name}, nMedals: #{@nMedals}, shieldPower: #{@shieldPower}, pendingDamage: [#{@pendingDamage}], weapons: {#{@weapons}}, shieldBoosters: {#{@shieldBoosters}}, hangar: [#{@hangar}]"
		end

        def assignFuelValue (f)
        	@fuelUnits = f <= @@MAXFUEL ? f : @@MAXFUEL
        end

		def cleanPendingDamage
			if @pendingDamage != nil or @pendingDamage.hasNoEffect
				@pendingDamage = nil
			end
		end
		
		def receiveWeapon(w) # receiveWeapon(w : Weapon) : boolean
			if @hangar != nil
				@hangar.addWeapon(w) # true si tiene éxito, false si no
			else
				false
			end
		end
		
		def receiveShieldBooster(s)
			if @hangar != nil
				@hangar.addShieldBooster(s) # true si tiene éxito, false si no
			else
				false
			end
		end
		
		def receiveHangar(h)
			if @hangar == nil 
				@hangar = h
			end
		end
		
		def discardHangar
			@hangar = nil
		end
		
		def receiveSupplies(s) # s: SuppliesPackage
			@fuelUnits += s.fuelUnits
			@shieldPower += s.shieldPower
			@ammoPower += s.ammoPower
		end
		
		def setPendingDamage(d) # setPendingDamage(d : Damage) : void
			@pendingDamage = d.adjust(@weapons, @shieldBoosters)
		end

		def mountWeapon (i) # mountWeapon(i : int) : void
			new_weapon = @hangar.weapons.delete_at(i) unless hangar.nil?
			@weapons << new_weapon unless new_weapon.nil?
		end

		def mountShieldBooster(i) # mountShieldBooster(i : int) : void
			new_shield_booster = @hangar.shieldBoosters.delete_at(i) unless hangar.nil?
			@shieldBoosters << new_shield_booster unless new_shield_booster.nil?
		end

		def discardWeaponInHangar (i)
			if @hangar != nil
				@hangar.removeWeapon(i)
			end
		end

		def discardShieldBoosterInHangar (i) 
			if @hangar != nil
				@hangar.removeShieldBooster(i)
			end
		end

		def getSpeed
			@fuelUnits / @@MAXFUEL
		end

		def move
			@fuelUnits = @fuelUnits >= getSpeed ? @fuelUnits - getSpeed : 0
		end

		def validState
			@pendingDamage == nil or @pendingDamage.hasNoEffect
		end

		def cleanUpMountedItems
			@weapons.delete_if {|x| x.uses == 0} unless @weapons.nil?
			@shieldBoosters.delete_if {|x| x.uses == 0} unless @shieldBoosters.nil?
		end

		# Especificación de acceso a métodos privados
		private :assignFuelValue, :cleanPendingDamage
    end
end

=begin
+discardShieldBooster(i : int) : void
+discardWeapon(i : int) : void
+fire() : float
+protection() : float
+receiveShot(shot : float) : ShotResult
+setLoot(loot : Loot) : void
=end

if $0 == __FILE__
	require_relative 'SuppliesPackage'
	require_relative 'Weapon'
	require_relative 'WeaponType'
	require_relative 'Hangar'
	require_relative 'Damage'

	my_space_station = Deepspace::SpaceStation.new("Nombre de mi estación", Deepspace::SuppliesPackage.new(27, 28, 29))
	puts my_space_station.to_s

	my_space_station.assignFuelValue(100)
	puts my_space_station.to_s

	puts
	d = Deepspace::Damage.newNumericWeapons(3,3)
	puts d.to_s
	my_space_station.setPendingDamage(d)
	puts my_space_station.to_s
	my_space_station.cleanPendingDamage
	puts my_space_station.to_s

	puts
	my_space_station.receiveHangar(Deepspace::Hangar.new(12))
	puts my_space_station.to_s
	my_space_station.receiveWeapon(Deepspace::Weapon.new(4, 5, 6))
	puts my_space_station.to_s
	my_space_station.receiveSupplies(Deepspace::SuppliesPackage.new(27, 28, 29))
	puts my_space_station.to_s

	puts
	my_space_station.setPendingDamage(Deepspace::Damage.newSpecificWeapons(Array.new(3){Deepspace::WeaponType::MISSILE}, 1))
	puts my_space_station.to_s

	puts "\nCosa que quiero comprobar\n\n"
	puts my_space_station.mountWeapon(0)
	puts my_space_station.to_s

	puts my_space_station.mountShieldBooster(3)
	puts my_space_station.to_s

	puts my_space_station.discardWeaponInHangar(0)
	puts my_space_station.to_s
	puts my_space_station.discardShieldBoosterInHangar(5)
	puts my_space_station.to_s

	puts my_space_station.getSpeed
	puts my_space_station.move
	puts my_space_station.validState
	my_space_station.cleanUpMountedItems
	puts my_space_station.to_s
end

File: P2/java/deepspace/src/deepspace/SpaceStation.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

public class SpaceStation {
    private static int MAXFUEL = 100;
    private static float SHIELDLOSSPERUNITSHOT = 0.1f;
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private float shieldPower;
    
    private Damage pendingDamage;
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;
    private Hangar hangar;
    
    SpaceStation(String n, SuppliesPackage supplies){
        name = n;
        ammoPower = supplies.getAmmoPower();
        fuelUnits = supplies.getFuelUnits();
        shieldPower = supplies.getShieldPower();
        
        weapons = new ArrayList<Weapon>();
        shieldBoosters = new ArrayList<ShieldBooster>();
    }
    
    public SpaceStationToUI getUIversion(){
        return new SpaceStationToUI(this);
    }
    
    public String toString(){   //NO SE SI ESTA BIEN
        return ("MAXFUEL: " + MAXFUEL + ", SHIELDLOSSPERUNITSHOT: " + SHIELDLOSSPERUNITSHOT + ", ammoPower: " + ammoPower + ", fuelUnits: " + fuelUnits + ", name: " + name.toString() + ", nMedals: " + nMedals + ", shieldPower: " + shieldPower + ", pendingDamage: [" + pendingDamage.toString() + "], weapons: {" + weapons.toString() + "}, shieldBooster: {" + shieldBoosters.toString() + "}, hangar: [" + hangar.toString() + "]");
    }
    
    private void assignFuelValue(float f){  // ES RARO QUE NO HAYAMOS USADO NI assignFuelValue NI cleanPendingDamage en toda la clase, deberiamos usarlas
        fuelUnits = f <= MAXFUEL ? f : MAXFUEL;
    }
    
    private void cleanPendingDamage(){
        if(pendingDamage.hasNoEffect())
            pendingDamage = null;
    }
    
    public float getAmmoPower(){
        return ammoPower;
    }
    
    public float getFuelUnits(){
        return fuelUnits;
    }
    
    public Hangar getHangar(){
        return hangar;
    }
    
    public String getName(){
        return name;
    }
    
    public int getNMedals(){
        return nMedals;
    }
    
    public Damage getPendingDamage(){
        return pendingDamage;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return shieldBoosters;
    }
    
    public float getShieldPower(){
        return shieldPower;
    }
    
    public ArrayList<Weapon> getWeapons(){
        return weapons;
    }
    
    public boolean receiveWeapon(Weapon w){
        if(hangar != null){
            return hangar.addWeapon(w);
        }else
            return false;
    }
    
    public boolean receiveShieldBooster(ShieldBooster s){
        if(hangar != null){
            return hangar.addShielBooster(s);
        }else
            return false;
    }
    
    public void receiveHangar(Hangar h){
        if(hangar == null){
            hangar = h;
        }
    }
    
    public void discardHangar(){
        hangar = null;
    }
    
    public void receiveSupplies(SuppliesPackage s){
        ammoPower += s.getAmmoPower();
        shieldPower += s.getShieldPower();
        fuelUnits += s.getFuelUnits();
    }
    
    public void setPendingDamage(Damage d){
        pendingDamage = d.adjust(weapons, shieldBoosters);
    }
    
    public void mountWeapon(int i){
        if(hangar != null){
            Weapon newWeapon = hangar.getWeapons().remove(i);   //No estoy seguro si remove devuelve el elemento eliminado
            if(newWeapon != null)
                weapons.add(newWeapon); 
        }
    }
    
    public void mountShieldBooster(int i){
        if(hangar != null){
            ShieldBooster newShield = hangar.getShieldBoosters().remove(i);   //No estoy seguro si remove devuelve el elemento eliminado
            if(newShield != null)
                shieldBoosters.add(newShield);
        }
    }
    
    public void discardWeaponInHangar(int i){
        if(hangar != null)
            hangar.removeWeapon(i);
    }
    
    public void discardShieldBoosterInHangar(int i){
        if(hangar != null)
            hangar.removeShieldBooster(i);
    }
    
    public float getSpeed(){
        return (float)(fuelUnits / MAXFUEL);
    }
    
    public void move(){
        fuelUnits = fuelUnits >= getSpeed() ? fuelUnits - getSpeed() : 0;
    }
    
    public boolean validState(){
        return (pendingDamage == null || pendingDamage.hasNoEffect());
    }
    
    public void cleanUpMountedItems(){
        if(weapons != null)
            for(int i=0; i<weapons.size(); ++i)
                if(weapons.get(i).getUses() == 0)
                    weapons.remove(i);
    
        if(shieldBoosters != null)
            for(int i=0; i<shieldBoosters.size(); ++i)
                if(shieldBoosters.get(i).getUses() == 0)
                    shieldBoosters.remove(i);
    }         
    
    // +discardWeapon(i : int) : void
    public void discardWeapon(int i){
        
    }
    
    // +discardShieldBooster(i : int) : void
    public void discardShieldBooster(int i){
        
    }
    
    /*
    +fire() : float
    +protection() : float
    +receiveShot(shot : float) : ShotResult
    +setLoot(loot : Loot) : void
    */
}
